### DISCUSSION TITLE
How to add database functions with DELIMITER

### DISCUSSION THREAD

**Discussion by tandibar** [prisma/prisma#12671] [Q&A]
I am trying to add database functions to my mysql database.

My first approach was via a seed script which works perfectly fine. But I think the functions do not belong to the seeds and should be included in the migrations. I copied the raw sql in a new migration after executing `prisma migrate dev --create-only`. This is a part of the SQL:
```sql
DELIMITER ;;
CREATE FUNCTION `gen_chars`(length INT) RETURNS varchar(255) CHARSET utf8
BEGIN
DECLARE identifier VARCHAR(255);
RETURN identifier;
      
END ;;
    
CREATE FUNCTION `gen_cuid`() RETURNS char(25) CHARSET utf8
BEGIN
    
DECLARE randoms CHAR(8);
    
return randoms;
    
END ;;
DELIMITER ;
```

If I execute `prisma migrate dev` I get the following error:

```bash
Error: P3018
A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve
Migration name: 20220405094924_initial_db_functions
Database error code: 1064
Database error:
You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'DELIMITER ;;
CREATE FUNCTION `gen_chars`(length INT) RETURNS varchar(255) CHA...' at line 1
Please check the query number 1 from the migration file. 
```

So it seems that the `DELIMITER` statement is a problem, but I do not understand why.

Can anyone explain to me how a add multiple database functions in one migration via prisma?

PS: I found a very similar question https://github.com/prisma/prisma/discussions/9611, but the solution is not my solution. But he seems to be using delimiter successfully.

**Comment by alex-semenyuk** [comment]
Did you try to use another delimiter, for example // (as it's suggested here https://mariadb.com/kb/en/delimiters/)? Maybe the issue is that semicolon standard delimiter.

**Reply by tandibar** [reply]
Good idea, but it also fails with the same error:
    
    Database error:                                                                                                                                                                             │
    You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'DELIMITER //
    CREATE FUNCTION `gen_chars`(length INT) RETURNS varchar(255) CHA...' at line 1

**Reply by alex-semenyuk** [reply]
@tandibar Ok, maybe problem similar with https://github.com/prisma/prisma/discussions/9611, but you use another keyword `length` as a name.

**Reply by tandibar** [reply]
I tested the following:

    DELIMITER //
    CREATE FUNCTION `gen_chars`() RETURNS varchar(255)
    BEGIN 
      RETURN "";
    END //
    DELIMITER ;

there should not be any reserved word in it but it still fails. If I remove the `DELIMITER` stuff like this:

    CREATE FUNCTION `gen_chars`() RETURNS varchar(255)
    BEGIN
      RETURN "";
    END

it works.

**Reply by alex-semenyuk** [reply]
Just to check - is not an issue white space between END and //

**Reply by tandibar** [reply]
I checked with 
    
    DELIMITER //
    CREATE FUNCTION `gen_chars`() RETURNS varchar(255)
    BEGIN 
      RETURN "";
    END//
    DELIMITER ;

still the same error.

**Comment by tandibar** [comment]
My workaround for now is to have one function per migration file. So instead of

    DELIMITER ;;
    CREATE FUNCTION `gen_chars`(length INT) RETURNS varchar(255) CHARSET utf8
    BEGIN
        DECLARE identifier VARCHAR(255);
        RETURN identifier;
    END ;;
        
    CREATE FUNCTION `gen_cuid`() RETURNS char(25) CHARSET utf8
    BEGIN
        DECLARE randoms CHAR(8);
        return randoms;
    END ;;
    DELIMITER ;

I have one file per function like this

    CREATE FUNCTION `gen_chars`(length INT) RETURNS varchar(255) CHARSET utf8
    BEGIN
        DECLARE identifier VARCHAR(255);
        RETURN identifier;
    END

and that    

    CREATE FUNCTION `gen_cuid`() RETURNS char(25) CHARSET utf8
    BEGIN
        DECLARE randoms CHAR(8);
        return randoms;
    END

I can no leave out the whole `DELIMITER` stuff. This is not a real solution to the problem.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
